### PR TITLE
AWS Glue import refactoring

### PR DESCRIPTION
**Pull request by roykoand**
- [x] Tests pass
- [x] ruff format
- [x] README.md updated (if relevant)
- [x] CHANGELOG.md entry added

### CODE REVIEW

File: datacontract/imports/glue_importer.py [hunk 1]
```diff
@@ -109,18 +109,17 @@ def get_glue_table_schema(database_name: str, table_name: str):
                     "Name": pk["Name"],
                     "Type": pk["Type"],
                     "Hive": True,
-                    "Comment": "Partition Key",
+                    "Comment": pk.get("Comment"),
```
Comment (roykoand) [inline datacontract/imports/glue_importer.py:112]: `"Comment": "Partition Key"` is overriding the column comment, I suggest to remove it and leave `required: true` as the only signal of the partition column

File: datacontract/imports/glue_importer.py [hunk 2]
```diff
@@ -253,30 +267,20 @@ def map_type_from_sql(sql_type: str) -> str:
...
-    if sql_type.startswith("date"):
-        return "date"
-    if sql_type.startswith("decimal"):
-        return "decimal"
-    return "variant"
+
+    type_mapping = {
+        "string": "string",
+        "int": "int",
+        "bigint": "bigint",
```
Comment (roykoand) [inline datacontract/imports/glue_importer.py:274]: Issue https://github.com/datacontract/datacontract-cli/issues/346 with `bigint` -> `long` conversion has been fixed. Also, some data types that are not used by Glue have been removed - `short`, `byte`, `text`. 
Also, `integer` has been renamed to `int` to be consistent with Glue (not backward compatible change).
Notice that `variant` was renamed to `unknown`, this is what Glue uses when something went wrong
image

File: datacontract/imports/glue_importer.py [hunk 3]
```diff
@@ -140,8 +139,13 @@ def import_glue(
     if table_names is None:
         table_names = get_glue_tables(source)
 
+    server_kwargs = {"type": "glue", "account": catalogid, "database": source}
+
+    if location_uri:
+        server_kwargs["location"] = location_uri
```
Comment (roykoand) [inline datacontract/imports/glue_importer.py:145]: `location_uri` with `"None"` string is considered to be None value while desearalizing to YAML.
Therefore Glue database without a location creates invalid data contract:
```
datacontract-cli git:(56f4dbe) ✗ datacontract test test1.yaml 
Testing test1.yaml
WARNING:root:Data Contract YAML is invalid. Validation error: data.servers.production must be valid exactly by one definition (0 matches found)
ERROR:root:Run operation failed: [lint] Check that data contract YAML is valid - None - failed - data.servers.production must be valid exactly by one definition (0 matches found) - datacontract
╭────────┬────────────────────────────────────────┬───────┬────────────────────────────────────────────────────╮
│ Result │ Check                                  │ Field │ Details                                            │
├────────┼────────────────────────────────────────┼───────┼────────────────────────────────────────────────────┤
│ failed │ Check that data contract YAML is valid │       │ data.servers.production must be valid exactly by   │
│        │                                        │       │ one definition (0 matches found)                   │
╰────────┴────────────────────────────────────────┴───────┴────────────────────────────────────────────────────╯
🔴 data contract is invalid, found the following errors:
1) data.servers.production must be valid exactly by one definition (0 matches found)
```
Let's ignore destination param if it's not specified

Comment (jochenchrist) [reply datacontract/imports/glue_importer.py:145]: Is the location optional in Glue?

Comment (roykoand) [reply datacontract/imports/glue_importer.py:145]: @jochenchrist wow, thanks for the fast feedback! Yes, it's optional.
That's the location description you get in their UI:
image
and confirming I do have some Glue databases without a location specified

File: datacontract/imports/glue_importer.py [hunk 4]
```diff
@@ -186,27 +184,43 @@ def create_typed_field(dtype: str) -> Field:
     """
     field = Field()
```
Comment (roykoand) [inline datacontract/imports/glue_importer.py:185]: Lots of Glue data types are not yet implemented:
- `binary`
- `char`
- `interval`
- `smallint`
- `tinyint`
- `union`
etc.

I am not quite sure how to add them cause the supported list of data types:
```
['number', 'decimal', 'numeric', 'int', 'integer', 'long', 'bigint', 'float', 'double', 'string', 'text', 'varchar', 'boolean', 'timestamp', 'timestamp_tz', 'timestamp_ntz', 'date', 'array', 'map', 'object', 
'record', 'struct', 'bytes', 'null']
```
https://datacontract.com/#data-types

Can I just extend this list of supported data types with the ones above? Thanks!

Comment (jochenchrist) [reply datacontract/imports/glue_importer.py:185]: Please try to map Glue Types to a logical data type (https://datacontract.com/#data-types
) that makes most sense. 

You can use a config object to specify the physical type:

```
type: bytes
config:
  glueType: binary
```

Some types (esp. union) may not be supported, add a warning here.

File: datacontract/imports/glue_importer.py (hunk 3 above)
Comment (jochenchrist) [reply datacontract/imports/glue_importer.py:145]: OK, do you still get this issue? `location` was already optional in the schema.
If you still have this error, can you post the server object that you get.

Comment (roykoand) [reply datacontract/imports/glue_importer.py:145]: @jochenchrist yup, it's optional in Pydantic model. However, the JSON schema of specification does not allow anything except URI string: https://github.com/datacontract/datacontract-specification/blob/816e6ae21e7ebdd2b9261734bd1a7ee47388574c/datacontract.schema.json#L261
Therefore, Glue database without location will create an invalid contract with `location: None`.
If you validate this contract with `datacontract test` you will get the error above

Comment (roykoand) [reply datacontract/imports/glue_importer.py:145]: @jochenchrist yup, it's optional in Pydantic model. However, the JSON schema of specification does not allow anything except URI string: https://github.com/datacontract/datacontract-specification/blob/816e6ae21e7ebdd2b9261734bd1a7ee47388574c/datacontract.schema.json#L446
Therefore, Glue database without location will create an invalid contract with `location: None`.
If you validate this contract with `datacontract test` you will get the error above.
We can make the

File: datacontract/imports/glue_importer.py (hunk 4 above)
Comment (roykoand) [reply datacontract/imports/glue_importer.py:185]: Amazing! Thank you, I will keep that in mind for the next PRs

File: datacontract/imports/glue_importer.py (hunk 3 above)
Comment (jochenchrist) [reply datacontract/imports/glue_importer.py:145]: If the location is not set, then the data contract YAML should be valid (location is not a required property):
```
servers:
  production:
    type: glue
    account: '123456789012'
    database: test_database
```

Comment (roykoand) [reply datacontract/imports/glue_importer.py:145]: @roykoand Yes, sorry for the confusion. It's fixed by this PR but was not working before. Thanks!

File: datacontract/imports/glue_importer.py (hunk 1 above)
Comment (jochenchrist) [reply datacontract/imports/glue_importer.py:112]: ok

Comment (roykoand) [reply datacontract/imports/glue_importer.py:112]: @jochenchrist Thanks for the review! Is there anything else that needs to be done in this PR?

**Comment by jochenchrist** [comment]
Thanks for your contribution. Happy to merge.

**Pull request merged**